node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeCandidate.java

```diff
@@ -159,76 +186,174 @@ private double skewWith(NodeResources resources) {
-            this.parent = Optional.of(parent);
-            return this;
+        @Override
+        public String toString() {
+            return node.id();
         }
 
-        Builder violatesSpares(boolean violatesSpares) {
-            this.violatesSpares = violatesSpares;
-            return this;
+    }
+
+    /** A candidate for which no actual node has been created yet */
+    static class VirtualNodeCandidate extends NodeCandidate {
+
+        /** The resources this node must have if created */
+        private final NodeResources resources;
+
+        /** Needed to construct the node */
+        private final LockedNodeList allNodes;
+        private final NodeRepository nodeRepository;
+
+        private VirtualNodeCandidate(NodeResources resources,
+                                     NodeResources freeParentCapacity,
+                                     Node parent,
+                                     boolean violatesSpares,
+                                     LockedNodeList allNodes,
+                                     NodeRepository nodeRepository) {
+            super(freeParentCapacity, Optional.of(parent), violatesSpares, false, true, false);
+            this.resources = resources;
+            this.allNodes = allNodes;
+            this.nodeRepository = nodeRepository;
         }
 
-        Builder surplusNode(boolean surplusNode) {
-            isSurplusNode = surplusNode;
-            return this;
+        @Override
+        public NodeResources resources() { return resources; }
+
+        @Override
+        public Optional<String> parentHostname() { return Optional.of(parent.get().hostname()); }
+
+        @Override
+        public NodeType type() { return NodeType.tenant; }
+
+        public Optional<Allocation> allocation() { return Optional.empty(); }
+
+        public Node.State state() { return Node.State.reserved; }
+
+        public boolean wantToRetire() { return false; }
+
+        public Flavor flavor() { return new Flavor(resources); }
+
+        public NodeCandidate allocate(ApplicationId owner, ClusterMembership membership, NodeResources requestedResources, Instant at) {
+            return withNode().allocate(owner, membership, requestedResources, at);
         }
 
-        Builder newNode(boolean newNode) {
-            isNewNode = newNode;
-            return this;
+        /** Called when the node described by this candidate must be created */
+        public NodeCandidate withNode() {
+            Optional<IP.Allocation> allocation;
+            try {
+                allocation = parent.get().ipConfig().pool().findAllocation(allNodes, nodeRepository.nameResolver());
+                if (allocation.isEmpty())
+                    throw new IllegalStateException("No free ip addresses on " + parent.get() + ": Cannot allocate node");
+            } catch (Exception e) {
+                throw new IllegalStateException("Failed allocating IP address on " + parent.get() +": ", e);
```

Comment:
The old code would just advance to the next node if resolving failing. If we get a node with broken IP, the allocation will basically not succeed until manual intervention since it would keep trying the same node? We should catch this and try next node if this fails.

Also, lets drop the stack trace for this, see #14599

Reply:
Done, please ptal. I think it would be better if we could expect this not to fail though - perhaps we should have a maintainer looking for it.

Reply:
It's the product of countless cases of DNS/IP misconfiguration upstream from us. It will continue to happen. 😞

Reply:
I would like to stop using DNS but not sure if that would avoid all of that?

Reply:
IP assignment and DNS issues typically correlate, at least in our internal infrastructure: If foo.example.com resolves to 1.1.1.1, but 1.1.1.1 resolves to bar.example.com, there's likely an IP conflict where two distinct hosts have been assigned the same IP.

Vespa itself also has strict hostname validation on startup, where hostname and IP is resolved and compared. That's another reason we do this check during allocation as the node would refuse to start in case of misconfiguration anyway.